src/analyz/mod.rs

```diff
@@ -95,37 +95,85 @@ fn vtable_descriptor_for_cast<'tcx>(
-        Some(x) => x,
-        // If there's no trait ref, it means the `Dynamic` predicates contain only auto traits.
-        // The vtable for this trait object is empty.
-        // TODO: handle this better (currently the output omits the "vtable" field)
-        None => return None,
-    };
-    Some(trait_ref)
+    fn instantiate_field_ty<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        field_def: &ty::FieldDef,
+        args: ty::GenericArgsRef<'tcx>
+    ) -> ty::Ty<'tcx> {
+        let field_unsubst_ty = tcx.type_of(field_def.did);
+        tcx.instantiate_and_normalize_erasing_regions(
+            args, ty::TypingEnv::fully_monomorphized(), field_unsubst_ty
+        )
+    }
+
+    match (*old_ty.kind(), *new_ty.kind()) {
+        // The two base cases: we are casting between two references or two raw
+        // pointers.
+        (ty::TyKind::Ref(_, old_pointee, _), ty::TyKind::Ref(_, new_pointee, _)) =>
+            vtable_descriptor_for_ref_or_ptr_cast(tcx, old_pointee, new_pointee),
+        (ty::TyKind::RawPtr(old_pointee, _), ty::TyKind::RawPtr(new_pointee, _)) =>
+            vtable_descriptor_for_ref_or_ptr_cast(tcx, old_pointee, new_pointee),
+
+        // It is also possible to cast between two arbitrary struct values,
+        // provided that each struct type differs in exactly one field (or a
+        // field within another struct field) that supports a vtable descriptor
+        // cast. (e.g., casting from Arc<Foo> to Arc<dyn Bar>).
+        (ty::TyKind::Adt(old_adt_def, old_args), ty::TyKind::Adt(new_adt_def, new_args))
+          if old_adt_def.is_struct() && new_adt_def.is_struct() => {
+            // Compute the fields from each struct value, and ensure that there
+            // are an equal number of fields as a sanity check.
+            let old_fields = &old_adt_def.non_enum_variant().fields;
+            let new_fields = &new_adt_def.non_enum_variant().fields;
+            if old_fields.len() != new_fields.len() {
+                return None;
+            }
+
+            // Recursively walk over the fields of each struct value, and for
+            // each pair of unequal fields, recursively compute their vtable
+            // descriptor cast.
+            let mut vtable: Option<ty::PolyTraitRef<'tcx>> = None;
+            let mut unequal_fields_count: usize = 0;
+            for (old_field_def, new_field_def) in old_fields.iter().zip(new_fields.iter()) {
+                let old_field_ty = instantiate_field_ty(tcx, old_field_def, old_args);
+                let new_field_ty = instantiate_field_ty(tcx, new_field_def, new_args);
+                if old_field_ty != new_field_ty {
+                    unequal_fields_count += 1;
+                    vtable = vtable_descriptor_for_cast(mir, kind, old_field_ty, new_field_ty);
+                }
+            }
+
+            // If we have found exactly one pair of unequal fields, then use the
+            // resulting vtable descriptor cast.
```

Comment:
Should note that we expect to find exactly one pair of unequal fields in all practical cases.  [Here's an counterexample](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=a7accacdd6e5f58c834528ddfeceeafa) that has two unequal fields, but (1) this seems to be special-cased to `PhantomData` only, and (2) it would be unusual to use `PhantomData` with `T` when `T` is mentioned elsewhere in the struct (there must be a non-phantom field like `*const T` somewhere so `CoerceUnsized` has somewhere to apply the coercion).

Reply:
Even though it's unusual, wouldn't we still want to support the `PhantomData` edge case, as in the first half of your linked example? I ask because `Arc` actually does hold two unequal fields - a `NonNull<ArcInner<T>>` and a `PhantomData<ArcInner<T>>` - and I think that in order to support casts of e.g. `Arc<Concrete>` to `Arc<dyn Trait>`, we need to accommodate this pattern.

Reply:
Wow! I didn't think there was any reason to use `PhantomData<T>` when `T` is already mentioned in the type, but apparently it's needed for dropck: https://github.com/rust-lang/rust/blob/f3662275e45271b6f0f1d30d572a21e19f3261a1/src/libcore/ptr.rs#L2480.  So we do need to handle `PhantomData` here after all.

Reply:
I've tried to address this in #132.